library/ssl_srv.c

```diff
@@ -1213,15 +1003,90 @@ static int ssl_parse_client_hello( mbedtls_ssl_context *ssl )
 
     buf = ssl->in_hdr;
 
+    /* Distinguish between SSLv2 and SSLv3+higher */
 #if defined(MBEDTLS_SSL_SRV_SUPPORT_SSLV2_CLIENT_HELLO)
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
     if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_STREAM )
 #endif
         if( ( buf[0] & 0x80 ) != 0 )
-            return( ssl_parse_client_hello_v2( ssl ) );
-#endif
+        {
+            /* SSLv2 */
+            size_t n;
+
+#if defined(MBEDTLS_SSL_RENEGOTIATION)
+            if( ssl->renego_status != MBEDTLS_SSL_INITIAL_HANDSHAKE )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "client hello v2 illegal for renegotiation" ) );
+                ssl->send_alert = MBEDTLS_SSL_ALERT_LEVEL_FATAL;
+                ssl->alert_type = MBEDTLS_SSL_ALERT_MSG_HANDSHAKE_FAILURE;
+                return( MBEDTLS_ERR_SSL_BAD_HS_CLIENT_HELLO );
+            }
+#endif /* MBEDTLS_SSL_RENEGOTIATION */
+
+            MBEDTLS_SSL_DEBUG_BUF( 4, "record header", buf, 5 );
+
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello v2, message type: %d",
+                                        buf[2] ) );
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello v2, message len.: %d",
+                                        ( ( buf[0] & 0x7F ) << 8 ) | buf[1] ) );
+            MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello v2, max. version: [%d:%d]",
+                                        buf[3], buf[4] ) );
+
+            /*
+             * SSLv2 Client Hello
+             *
+             * Record layer:
+             *     0  .   1   message length
+             *
+             * SSL layer:
+             *     2  .   2   message type
+             *     3  .   4   protocol version
+             */
+            if( buf[2] != MBEDTLS_SSL_HS_CLIENT_HELLO ||
+                buf[3] != MBEDTLS_SSL_MAJOR_VERSION_3 )
+            {
+                MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad client hello message" ) );
+                return( MBEDTLS_ERR_SSL_BAD_HS_CLIENT_HELLO );
+            }
 
-    MBEDTLS_SSL_DEBUG_BUF( 4, "record header", buf, mbedtls_ssl_hdr_len( ssl ) );
+            n = ( ( buf[0] << 8 ) | buf[1] ) & 0x7FFF;
+
+            if( n < 17 || n > 512 )
```

Comment:
Magic numbers

Reply:
While I agree in principle, I prefer not to change the actual parsing functions at the moment more than necessary.